.PHONY: install trim_galore deconseq fastq_pair megahit spades download prodigal hmmer busco metaquast fastani micro reads nextflow java run log clean

help:
	@echo "make nextflow --> install nextflow."
	@echo "make reads --> Make files.csv from local input files."
	@echo "make install --> Install docker images."
	@echo "make cloud --> Run the pipeline on the cloud via AWS BATCH."
	@echo "make local --> Run the pipeline locally."
	@echo "make log --> Get log information for a particular session."
	@echo "make clean --> Clean the work directory generated by nextflow."


install: trim_galore deconseq fastq_pair megahit prodigal download hmmer busco metaquast
	@echo "Installing the required docker images..."

trim_galore:
	# Trim_galore
	# This is how you run it - docker run --rm bschiffthaler/trim-galore:0.6.5 -help
	@echo "Pulling trim_galore for read quality trimming and assessment"
	docker pull bschiffthaler/trim-galore:0.6.5

deconseq:
	# Deconseq
	@echo "Pulling deconseq for removing potential contaminants..."
	docker pull quay.io/grbot/deconseq:latest
	wget https://sourceforge.net/projects/deconseq/files/standalone/deconseq-standalone-0.4.3.tar.gz --no-check-certificate

fastq_pair:
	# fastq_pair
	@echo "pulling fastq_pair for syncing paired-end reads and removing singletons.."
	docker pull spashleyfu/debian_stretch_fastq_pair:v1.0

megahit:
	# Megahit
	@echo "Pulling megahit for metagenome assembly"
	docker pull nanozoo/megahit:1.2.9--87c4487
	
spades:
	# Spades
	@echo "Pulling spades for metagenome assembly..."
	docker pull staphb/spades:3.15.5

prodigal:
	# Prodigal
	@echo "Pulling prodigal for ORF prediction..."
	docker pull biocontainers/prodigal:v1-2.6.3-4-deb_cv1

download:
	@echo "Downloading PFAM HMM database..."
	wget https://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/Pfam-A.hmm.gz

hmmer:
	# Hmmer
	@echo "Pulling Hmmer for performing hmm database searches"
	docker pull staphb/hmmer:3.3.2
	
busco:
	#BUSCO for assessing assembly completeness
	@echo "Pulling BUSCO for assembly completeness assessment...."
	docker pull staphb/busco:5.4.7
	
metaquast:
	#Metaquast
	@echo "Pulling metaquast for assembly contiguity assessment...."
	docker pull cami/metaquast:latest

fastani:
	# Fast ani for genenome to genome comparisons
	@echo "Pulling Fast ani for genenome to genome comparisons"
	docker pull staphb/fastani:1.34-rgdv2

micro:
	# Microbiology image
	# the executables are locate /software/Miniconda3/bin/
	# see the docker page here https://hub.docker.com/r/fanyucai1/micro
	# for the full list of software installed
	# This is how to run the software - docker run --rm fanyucai1/micro:latest /software/Miniconda3/bin/megahit -h
	@echo "Pulling micro, a huge docker image with many useful microbialogy tools installed"
	docker pull fanyucai1/micro:latest

reads:
	@echo "Making files.csv from local input files with sample names as directory names in a folder named 01.raw_data/"
	bash make_files.csv.sh

nextflow: java
	@echo "Installing Nextflow..."
	wget -O nextflow https://www.nextflow.io/releases/latest/nextflow?a=update && chmod +x nextflow
	./nextflow
	sudo mv ./nextflow /usr/local/bin/
java:
	@echo "Installing Java for nextflow"
	sudo yum update -y && sudo yum install -y java && sudo systemctl start docker

cloud:
	@echo "Running the pipeline on the cloud via AWS BATCH"
	nextflow -C nextflow.config run main.nf -resume -profile cloud

local:
	@echo "Running the pipeline locally"
	nextflow -C nextflow.config run main.nf -resume -profile local

log:
	@echo "Get log information for a session named desperate_dijkstr"
	@echo "The get the sesson name type:"
	@echo "nextflow log"
	@echo "at the command line."
	nextflow log desperate_dijkstra -f 'name,process,workdir,duration,script,status'
clean:
	@echo "Cleaning up the work directory generated by nextflow.."
	nextflow clean